File: src/main/java/com/sstep/recipe/manager/dao/RecipeDao.java
package com.sstep.recipe.manager.dao;


import com.sstep.recipe.manager.entity.Recipe;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.List;


@Repository
public class RecipeDao {

    @PersistenceContext
    private EntityManager em;

    public List<Recipe> findAll() {
        return em.createQuery("SELECT recipe FROM Recipe recipe", Recipe.class)
                .getResultList();
    }

    public Recipe findById(final long id) {
        return em.createQuery("SELECT recipe FROM Recipe recipe WHERE recipe.id = :id", Recipe.class)
                .setParameter("id", id)
                .getSingleResult();
    }

    public Recipe save(final Recipe recipe) {
        em.persist(recipe);
        return recipe;
    }
}

File: src/main/kotlin/com/sstep/recipe/manager/dao/RecipeDao.kt
package com.sstep.recipe.manager.dao

import com.sstep.recipe.manager.entity.Recipe
import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
import org.springframework.stereotype.Repository

@Repository
open class RecipeDao {

    @PersistenceContext
    private lateinit var em: EntityManager

    open fun findAll(): List<Recipe> {
        return em.createQuery("SELECT recipe FROM Recipe recipe", Recipe::class.java)
            .resultList
    }

    open fun findById(id: Long): Recipe {
        return em.createQuery("SELECT recipe FROM Recipe recipe WHERE recipe.id = :id", Recipe::class.java)
            .setParameter("id", id)
            .singleResult
    }

    open fun save(recipe: Recipe): Recipe {
        em.persist(recipe)
        return recipe
    }

    open fun remove(id: Long) {
        val recipe = findById(id)
        em.remove(recipe)
    }
}